app/src/test/java/org/simple/clinic/user/UserSessionTest.kt

```diff
@@ -127,53 +132,110 @@ class UserSessionTest {
+    )
+  }
 
-    val result3 = userSession.loginWithOtp("").blockingGet()
-    assertThat(result3).isInstanceOf(LoginResult.ServerError::class.java)
+  @Test
+  @Parameters(method = "parameters for triggering sync on login")
+  fun `sync should be triggered on successful login`(
+      response: Single<LoginResponse>,
+      shouldTriggerSync: Boolean
+  ) {
+    var syncInvoked = false
+
+    whenever(loginApi.login(any())).thenReturn(response)
+    whenever(syncScheduler.syncImmediately()).thenReturn(Completable.complete().doOnComplete { syncInvoked = true })
 
-    val result4 = userSession.loginWithOtp("").blockingGet()
-    assertThat(result4).isInstanceOf(LoginResult.NetworkError::class.java)
+    userSession.loginWithOtp("000000").blockingGet()
+
+    if (shouldTriggerSync) {
+      assertThat(syncInvoked).isTrue()
+    } else {
+      verifyZeroInteractions(syncScheduler)
+    }
   }
 
-  private fun <T> unauthorizedHttpError(): HttpException {
-    val error = Response.error<T>(401, ResponseBody.create(MediaType.parse("text"), unauthorizedErrorResponseJson))
-    return HttpException(error)
+  @Suppress("Unused")
+  private fun `parameters for triggering sync on login`(): List<List<Any>> {
+    return listOf(
+        listOf(Single.just(LoginResponse("accessToken", loggedInUserPayload)), true),
+        listOf(Single.error<LoginResponse>(NullPointerException()), false),
+        listOf(Single.error<LoginResponse>(unauthorizedHttpError<LoginResponse>()), false),
+        listOf(Single.error<LoginResponse>(IOException()), false)
+    )
   }
 
   @Test
-  fun `when find existing user then the network response should correctly be mapped to results`() {
-    val notFoundHttpError = mock<HttpException>()
-    whenever(notFoundHttpError.code()).thenReturn(404)
+  @Parameters(value = [
+    "true",
+    "false"
+  ])
+  fun `error in sync should not affect login result`(syncWillFail: Boolean) {
+    whenever(loginApi.login(any())).thenReturn(Single.just(LoginResponse("accessToken", PatientMocker.loggedInUserPayload())))
+    whenever(syncScheduler.syncImmediately()).thenAnswer {
+      if (syncWillFail) Completable.error(RuntimeException()) else Completable.complete()
+    }
+
+    var thrownException: Throwable? = null
 
-    whenever(registrationApi.findUser("123")).thenReturn(Single.just(PatientMocker.loggedInUserPayload()))
-    whenever(registrationApi.findUser("456")).thenReturn(Single.error(SocketTimeoutException()))
-    whenever(registrationApi.findUser("789")).thenReturn(Single.error(notFoundHttpError))
-    whenever(registrationApi.findUser("000")).thenReturn(Single.error(NullPointerException()))
+    try {
+      userSession.loginWithOtp("000000").blockingGet()
+    } catch (e: Throwable) {
+      thrownException = e
+    }
 
-    val result1 = userSession.findExistingUser("123").blockingGet()
-    assertThat(result1).isInstanceOf(FindUserResult.Found::class.java)
+    assertThat(thrownException).isNull()
```

Comment:
JUnit offers a slightly better way of expecting exceptions with `ExpectedException` rule. Do you want to use that? 

```kotlin
@Rule
val expectedException: ExpectedException = ExpectedException.none()

@Test
fun foo() {
  expectedException.expect(AnyException.class)
  expectedException.expectMessage("The expected message")
}
```

Reply:
Oh wait, this test is incorrect. I need to check the result of the `loginWithOtp` call and it should be `Success` in all cases. Thanks for calling attention to this!

Reply:
Fixed.